### PR TITLE
Psvmi integration attempt #2

### PR DESCRIPTION
**Pull request by sahilsuneja1**

### CODE REVIEW

**Comment by MHBauer** [comment]
Does this contain the fixes for #122? Do you need help squashing?

**Comment by sahilsuneja1** [comment]
This replaces #122. Thanks for the offer!

**Comment by sahilsuneja1** [comment]
Hi, I've updated .travis.yml as well, and the build is passing there. So I guess this is good to merge. (already reviewed by @ricarkol ). It might need some squashing now, not touching it for fear of breaking!

File: .travis.yml [hunk 1]
```diff
@@ -26,9 +26,18 @@ before_install:
     # might as well upgrade pip to support TLS and get rid of the warnings
     - sudo -H pip install --upgrade pip
     - sudo -H pip install codecov pytest-cov
+    # psvmi specific
+    - sudo apt-get install qemu-kvm
+    - rm -rf psvmi
```
Comment (MHBauer) [inline .travis.yml]: where is this directory coming from that we have to remove it?

File: Dockerfile.test [hunk 2]
```diff
@@ -1,19 +1,21 @@
-FROM alpine:3.3
+FROM ubuntu:14.04
```
Comment (MHBauer) [inline Dockerfile.test]: this image is 100x larger, so this will take longer. Can we use 16.04? @ricarkol

**Comment by MHBauer** [comment]
Squish all the travis changes at least.

**Comment by MHBauer** [comment]
flake8 warnings up from 516 -> 846. It would be really great to not add 300+ nits.

File: tests/unit/test_features_crawler.py [hunk 3]
```diff
@@ -514,26 +514,29 @@ def test_crawl_os_outvm_mode_without_vm_failure(self, *args):
...
+                                                       100000,
+                                                       100000))
     def test_crawl_os_outvm_mode_without_vm(self, *args):
         fc = FeaturesCrawler(crawl_mode=Modes.OUTVM,
                              vm=('dn', '2.6', 'ubuntu', 'x86'))
         for os in fc.crawl_os():
             pass
         assert args[0].call_count == 1
-        args[0].assert_called_with('dn', '2.6', 'ubuntu', 'x86')
+        #args[0].assert_called_with('dn', '2.6', 'ubuntu', 'x86')
```
Comment (MHBauer) [inline tests/unit/test_features_crawler.py]: do we need this?

File: tests/functional/test_features_crawler_OUTVM.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+
+# Tests the FeaturesCrawler class
+# Throws an AssertionError if any test fails
+
+class FeaturesCrawlerTests(unittest.TestCase):
+   
+    USE_RUNNING_VM = False
+    # USE_RUNNING_VM = True requires --pid=host on docker run command line in Makefile
+    #vmIDs = [['vm2','3.2.0-101-generic_3.2.0-101.x86_64','ubuntu','x86_64']]
```
Comment (MHBauer) [inline tests/functional/test_features_crawler_OUTVM.py]: can we get rid of this line?

File: tests/functional/test_features_crawler_OUTVM.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+    
+    SETUP_ONCE = False
+
+    crawlers = [] 
+    
+    #vmIDs = [['vm2', '4.0.3.x86_64', 'vanilla', 'x86_64']]
```
Comment (MHBauer) [inline tests/functional/test_features_crawler_OUTVM.py]: can we get rid of this line?

File: Dockerfile.test [hunk 6]
```diff
@@ -23,8 +25,14 @@ WORKDIR /crawler
...
+COPY psvmi/maps maps
+COPY psvmi/offsets offsets
+COPY psvmi/header.h .
+
 CMD (docker daemon > ../docker.out 2>&1 &) && \
 	sleep 5 && \
         flake8 --max-complexity 10 . || true && \
         pylint crawler || true && \
-        python setup.py test --addopts '--cov=.'
+        python setup.py test --addopts '-s --cov=.'
```
Comment (MHBauer) [inline Dockerfile.test]: are the tests printing a lot of stuff that we need to turn off?

File: tests/functional/test_features_crawler_OUTVM.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+            for item in crawler.crawl_cpuHw():
+                assert 'QEMU' in str(item)
+
+    
+    def test_features_crawler_crawl_outvm_process(self):
+        for crawler in self.crawlers:
+            output = crawler.crawl_processes()
+            assert len(list(output))>0
+            
+            #assert any('bash' in ProcessFeature._make(item[1]).pname for item in crawler.crawl_processes())
```
Comment (MHBauer) [inline tests/functional/test_features_crawler_OUTVM.py]: do we need this?

**Comment by MHBauer** [comment]
Comments on some of the new functions would be nice.

**Comment by sahilsuneja1** [comment]
Thanks @MHBauer , I'll fix these soon with the next commit.

File: Dockerfile.test (hunk 6 above)
Comment (sahilsuneja1) [inline Dockerfile.test]: '-s' allows verbosity. Original was silent. I only added -s to see output for my test code. Other tests's do print lots of stuff. So off / on (-s) ?

File: Dockerfile.test (hunk 2 above)
Comment (sahilsuneja1) [inline Dockerfile.test]: I'll test with 16.04. I had to move away from alpine due to segfaults in its libc (musl), arising from possibly inefficient  C + python wrapping in psvmi. I'll flag this as in 'issue'

**Comment by sahilsuneja1** [comment]
Okay, suggested fixes mostly applied. Merge request please.

**Comment by ricarkol** [comment]
Hey Sahil, if you give me permissions on your fork, i can squash the commits and push it there. If not, try this:
```
git rebase -i HEAD~7 # 7 is the number of commits you have to squash
```
that will open a text editor. Just change edit to squash on the last 6 commits. Save and exit. This will open a text editor again for you to edit the commit message for the new big commit. It defaults to all commit messages one after the other. Just pick one of them and leave the `signed by` line. Then force push your branch: `git push -f xxx`.

**Comment by sahilsuneja1** [comment]
@ricarkol: I already squashed several intermediate ones. I've added you as 'collaborator',would that give you the necessary permissions?

**Comment by ricarkol** [comment]
OK @sahilsuneja1 squashed into a single commit. If you want to get this commit back to your local branch you have to do the following (a pull won't work):
```
git checkout psvmi-integration
git fetch sahil
git reset sahil/psvmi-integration --hard
```
all of this assumes that `sahil` is the name of your remote repo (the one in github).

**Comment by ricarkol** [comment]
Unit tests are failing though:

```
tests/unit/test_container.py .....
tests/unit/test_dockercontainer.py ...................
tests/unit/test_dockerutils.py ...........................
tests/unit/test_emitter.py ...............................
tests/unit/test_features_crawler.py .........................................F...........F............
tests/unit/test_misc.py ...........
tests/unit/test_mtgraphite.py ....
tests/unit/test_namespace.py .............
tests/unit/test_package_utils.py ...
```

**Comment by ricarkol** [comment]
Went through the branch with Sahil. We decided to leave the unit tests for psvmi to next week, tracked here: https://github.com/cloudviz/agentless-system-crawler/issues/128. Rebased on top of the latest master.

**Comment by sahilsuneja1** [comment]
@MHBauer : addressed almost all comments, except flake- autopep reduced 150 but don't know how to bring it down any further.

**Comment by ricarkol** [comment]
@MHBauer i'm going to take a detailed pass to make pep, flake, and pylint as happy as possible after we get to 90% coverage (we are at 88%). This change looks good to me. If you are ok with it, can we merge it?

**Comment by sahilsuneja1** [comment]
Hey guys, for clarity, this has been rebased to the latest master and passes on Travis: https://travis-ci.org/sahilsuneja1/agentless-system-crawler. So, its good to go.

**Comment by MHBauer** [comment]
Looks good. We can get everything else later.

**Comment by ricarkol** [comment]
"We can get everything else later." sounds like famous last words. But I promise I'm taking a pass to make flake happy. Maybe I'm being naive, but I'm hoping it won't require too much refactoring.

**Pull request merged**